src/librustc/mir/interpret/queries.rs

```diff
@@ -2,24 +2,46 @@ use super::{ConstEvalResult, ErrorHandled, GlobalId};
 
 use crate::mir;
 use crate::ty::subst::{InternalSubsts, SubstsRef};
-use crate::ty::{self, TyCtxt};
+use crate::ty::{self, Ty, TyCtxt};
 use rustc_hir::def_id::DefId;
 use rustc_span::Span;
 
 impl<'tcx> TyCtxt<'tcx> {
     /// Evaluates a constant without providing any substitutions. This is useful to evaluate consts
     /// that can't take any generic arguments like statics, const items or enum discriminants. If a
-    /// generic parameter is used within the constant `ErrorHandled::ToGeneric` will be returned.
+    /// generic parameter is used within the constant `ErrorHandled::TooGeneric` will be returned.
     pub fn const_eval_poly(self, def_id: DefId) -> ConstEvalResult<'tcx> {
-        // In some situations def_id will have substitutions within scope, but they aren't allowed
-        // to be used. So we can't use `Instance::mono`, instead we feed unresolved substitutions
-        // into `const_eval` which will return `ErrorHandled::ToGeneric` if any of them are
-        // encountered.
-        let substs = InternalSubsts::identity_for_item(self, def_id);
-        let instance = ty::Instance::new(def_id, substs);
-        let cid = GlobalId { instance, promoted: None };
-        let param_env = self.param_env(def_id).with_reveal_all();
-        self.const_eval_global_id(param_env, cid, None)
+        let ty = self.type_of(def_id);
+        debug!("const_eval_poly: running on {:?}: {:?}", def_id, ty);
+
+        // FIXME: This check could be smarter / faster.
+        let is_generic = |t: Ty<'_>| -> bool {
+            for ty in t.walk() {
+                if let ty::Param(_) = ty.kind {
+                    debug!("{:?} is generic", t);
+                    return true;
+                }
+            }
+            debug!("{:?} is not generic", t);
+            false
+        };
+
+        if is_generic(ty) {
+            // If we don't necessarily know what the type is that we're dealing with, to begin
+            // with, then we're in too generic of a context. One place this happens is in the
+            // UnusedBrokenConst lint, when visiting an associated const for a blanket impl.
+            Err(ErrorHandled::TooGeneric)
+        } else {
+            // In some situations def_id will have substitutions within scope, but they aren't allowed
+            // to be used. So we can't use `Instance::mono`, instead we feed unresolved substitutions
+            // into `const_eval` which will return `ErrorHandled::ToGeneric` if any of them are
+            // encountered.
+            let substs = InternalSubsts::identity_for_item(self, def_id);
+            let instance = ty::Instance::new(def_id, substs);
+            let cid = GlobalId { instance, promoted: None };
+            let param_env = self.param_env(def_id).with_reveal_all();
```

Comment:
right here ,if you skip doing the reveal all if the instance `.needs_subst()` you should hopefully not get an  ICE anymore, because the relevant code bails out with TooGeneric

Reply:
Apologies if I wasn't clear -- I haven't been getting an ICE. In my previous comment, by an error bubbling up, I meant an Err being returned upwards (which ends in us gracefully emitting a diagnostic to the user). Hmm -- I'm wonder if doing this anyway will bail out with TooGeneric where I need it to.

Reply:
yes, without reveal all, such errors will not be emitted but become TooGeneric

Reply:
This hasn't been addressed.

Reply:
Most feedback hasn't yet. I'm still working on this pr. I think it's showing as outdated because I squashed & dropped some commits. I will ping you all and re-mark as waiting on review when this is ready to go again 👍